#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data_path = 'drive/MyDrive/Data_analitic/Python/Pandas_Python/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [9]:
pd.pivot_table(df, index=['City','Branch'], values='gross income', aggfunc='sum').sort_values(by='gross income', ascending=False)

gross income
City      Branch              
Naypyitaw C          5265.1765
Yangon    A          5057.1605
Mandalay  B          5057.0320

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [14]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', aggfunc='size')

In [15]:
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [34]:
top_branch = invoices_by_category['Electronic accessories'].idxmax()
top_branch_1 = invoices_by_category['Electronic accessories'].max()

In [37]:
print(f'Найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" у філіалі {top_branch} магазину Branch = {top_branch_1}')

Найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" у філіалі A магазину Branch = 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


#1. Як гендер та тип клієнта впливають на вибір товарних категорій та середній чек?

In [15]:
pd.pivot_table(df, index='Product line', columns=['Gender', 'Customer type'], values='Total', aggfunc='mean').round(2)

Gender                  Female            Male        
Customer type           Member  Normal  Member  Normal
Product line                                          
Electronic accessories  318.67  325.48  310.35  323.04
Fashion accessories     321.76  312.55  287.21  294.59
Food and beverages      384.29  348.00  273.46  273.54
Health and beauty       279.31  299.46  405.85  292.91
Home and lifestyle      366.69  400.08  298.44  290.70
Sports and travel       326.63  322.07  321.56  356.48

###Категорія: Health and beauty

Чоловіки – Member: 405.85
Жінки – Member: 279.31
#####Чоловіки з типом клієнта Member витрачають суттєво більше на товари категорії Health and beauty, ніж жінки. Це може свідчити про:

- вищу зацікавленість чоловіків у доглядових товарах, особливо якщо вони отримують сповіщення про акції через програму лояльності.
Купівлю подарунків для партнерок — акції в рамках програми можуть стимулювати такі покупки.
Регулярне придбання базових доглядових засобів — наприклад, гелі для душу, засоби для гоління, шампуні.
-віддають перевагу знайомому супермаркету, де їх усе влаштовує — розташування, асортимент, швидкість обслуговування.
- менш схильні вивчати альтернативи або "просто зайти подивитися".
Якщо вони вже стали учасниками програми лояльності — залишаються в ній і не шукають додаткових акцій або інших точок покупки.

- жінки, ймовірно, менше купують косметику в супермаркетах, натомість надають перевагу:
спеціалізованим сайтам (з ширшим асортиментом),
брендовим магазинам,
рекомендаціям косметологів.

###Категорія: Home and lifestyle

Жінки – Normal: 400.08
Чоловіки – Normal: 290.70
#####Жінки з типом клієнта Normal демонструють найвищі середні витрати в категорії Home and lifestyle. Це може свідчити про:

- частішу участь жінок у придбанні товарів для дому, навіть без участі в програмі лояльності.
- регулярність покупок, незалежно від акцій чи бонусів — можливо, мова йде про звичні потреби (посуд, побутові дрібниці, текстиль тощо).
- менша чутливість до стимулювання через програму лояльності в цій категорії — жінки просто приходять і купують те, що потрібно, без додаткових мотивацій.
- більш схильні до імпульсивних покупок, особливо у візуально привабливих зонах магазину (свічки, текстиль, декор).
- домашні товари часто викликають емоційний відгук ("гарна чашка", "затишна ковдра") — саме такий тип товарів сприяє незапланованим витратам.

####Рекомендації:
- запровадити цільові акції саме для Normal-жінок — наприклад, "3+1", подарунок за покупку, знижки на регулярні позиції.
- пропонувати участь у програмі лояльності під час покупок у категорії Home and lifestyle, щоб збільшити повторні візити.
- виділити категорію у візуальному оформленні супермаркету (наприклад, тематичні зони "Твій затишний дім").
- утримання важливіше за залучення: для чоловіків варто зосередитись на підвищенні зручності, швидкості та стабільності обслуговування.
- програма лояльності для чоловіків має бути простою, зрозумілою і не вимагати додаткових зусиль — наприклад, автоматичні бонуси або знижки без накопичень.
- оформлення магазинів і комунікації мають бути лаконічними — менше емоційного маркетингу, більше чіткої користі.

#2. Як змінюється середній чек залежно від способу оплати в різних містах?

In [16]:
pd.pivot_table(df, index='City', columns='Payment', values='Total', aggfunc='mean').round(2)

Payment      Cash  Credit card  Ewallet
City                                   
Mandalay   321.27       342.61   296.58
Naypyitaw  347.47       309.46   350.52
Yangon     307.10       318.22   312.10

###Загальна тенденція:
У більшості міст готівка (Cash) демонструє вищий середній чек — 326.18+.
Найнижчий середній чек по всім способам оплати — Ewallet у Mandalay (296.58).

- Naypyitaw: Найвищий середній чек у користувачів Ewallet (350.52) — що є нетиповим серед усіх міст. Можливо, електронні гаманці використовуються для великих покупок через акції/кешбеки.
А, можливо:
- цифрова зрілість населення:
Naypyitaw — столиця, адміністративний центр, можливо вища частка молодого, технічно підкованого населення.
- активніше використовують сучасні фінансові інструменти.
Акції, бонуси або кешбек на Ewallet. Ймовірно, магазин стимулює оплату електронними гаманцями (знижки, бонуси, “1% назад” тощо). Покупці накопичують витрати в один платіж, щоб отримати кешбек.
- великий сегмент “преміальних” покупців. Електронні гаманці можуть використовуватися людьми з вищими доходами, які купують дорожчі товари.
- Mandalay: Найбільший середній чек — при оплаті кредитною карткою (342.61). Це може вказувати на високу лояльність до банківських продуктів або часте використання кредитних лімітів.
- Yangon: Найменший розкид між способами оплати — приблизно однакові середні чеки. Це свідчить про більш рівномірне використання платіжних засобів.

####Рекомендації:
- для Naypyitaw: посилити акції на оплату Ewallet, адже користувачі цієї опції вже демонструють високу купівельну спроможність.
- для Mandalay: розвивати партнерства з банками — кредитні картки тут приносять найвищі суми. Можна впровадити спеціальні пропозиції для оплати саме карткою.
- для Yangon: нейтральна поведінка вказує на стабільність — тут можна тестувати нові методи оплати без ризику сильного відхилення в середньому чеку.

3. Які категорії товарів більше купують чоловіки, а які — жінки?

In [17]:
pd.pivot_table(df, index='Product line', columns='Gender', aggfunc='size')

Gender                  Female  Male
Product line                        
Electronic accessories      84    86
Fashion accessories         96    82
Food and beverages          90    84
Health and beauty           64    88
Home and lifestyle          79    81
Sports and travel           88    78

####Висновки:
Жінки купують більше у категоріях Fashion accessories, Food and beverages, Sports and travel — це відповідає типовим уявленням про інтереси.
Чоловіки випереджають жінок у категоріях Health and beauty та Electronic accessories — цікава тенденція, особливо в Health and beauty.
За Electronic accessories чоловіки мають трохи більшу кількість покупок, що логічно, враховуючи інтереси.
####Рекомендації:
Для Health and beauty розглянути розширення асортименту і спеціальні пропозиції, орієнтовані на чоловіків (наприклад, чоловіча косметика, засоби догляду).
Водночас варто збільшити залученість жінок — можливо, через окремі промоакції, знижки або нові товари, які зацікавлять саме жіночу аудиторію.
#####Розглянути фізичне розмежування полиць з чоловічими та жіночими товарами, щоб створити комфортнішу атмосферу для покупців обох статей. Це може допомогти уникнути ситуацій, коли хтось соромиться купувати певні товари, коли поруч представники іншої статі.
######Організувати маркетингові кампанії, які підкреслюватимуть різноманітність і орієнтацію на індивідуальні потреби кожної групи.
Для категорій, де жінки купують більше (Fashion accessories, Food and beverages, Sports and travel), підтримувати та розвивати асортимент, можливо, вводити додаткові акції.
Розробити цільові маркетингові кампанії для кожної статі, базуючись на їхніх перевагах у категоріях.

##Додаткові питання:
- У якому місті середній чек у жінок більший, а в якому — у чоловіків?
- В якому місті та філії більше купують — жінки чи чоловіки?
- Який товар у кожній категорії найдорожчий/найдешевший і скільки разів його купили?
- Який вплив має час покупки (Hour) на вибір категорії товарів?
- Що таке рейтинг (Rating), чи є погані, як вони впливають на поведінку?
- У якому місті більше покупців за гендером і типом клієнта?

Відповіді на деякі додаткові питання

In [ ]:
pd.pivot_table(df,index='Product line', columns='Gender', values='gross income', aggfunc='sum').round (2)

айбільший прибуток від жінок приносить категорія Food and beverages — 1579.57.
Найбільший прибуток від чоловіків — категорія Health and beauty — 1458.70.

####Жінки:

Витрачають найбільше в категоріях, пов'язаних із регулярним побутовим споживанням (Food and beverages, Fashion accessories, Home and lifestyle).
Це може бути пов’язано з тим, що жінки частіше відповідають за домашні покупки, планування продуктового кошика та створення побутового комфорту.

####Чоловіки:

Найбільші витрати спостерігаються в категорії Health and beauty — це може бути пов'язано з доглядовими засобами для себе або з купівлею подарунків.
Також мають стабільні витрати в Electronic accessories і Sports and travel.


####Рекомендації для бізнесу:
Food and beverages (для жінок):
- запустити акції “2+1” або крос-продажі (напр., кава + печиво).
- виділити товари, які часто купують жінки, в окремі зони.

Health and beauty (для чоловіків):
- розширити чоловічий асортимент.
- створити прості набори догляду (гоління, душ, догляд за тілом) — вигідно для самостійних або зайнятих покупців.

Fashion & Lifestyle (для жінок):
- використати візуальний мерчендайзинг — оформлення стендів, подарункові набори.
- пропонувати бонуси або знижки для власників карток лояльності.

Цільовий маркетинг:
Використовувати рекламу з урахуванням гендерних вподобань: email-розсилки, push-сповіщення з персоналізованими пропозиціями.

In [18]:
pd.pivot_table(df,index=['Customer type','City'], columns='Gender', aggfunc='size')

Gender                   Female  Male
Customer type City                   
Member        Mandalay       85    80
              Naypyitaw      96    73
              Yangon         80    87
Normal        Mandalay       77    90
              Naypyitaw      82    77
              Yangon         81    92

- Загалом гендерного дисбалансу у виборі типу покупця немає — кількість членів програми лояльності приблизно однакова серед чоловіків і жінок (501 проти 499).
- У Naypyitaw значно більше жінок-учасниць програми лояльності (96 проти 73 чоловіків).
- У Mandalay спостерігається протилежна ситуація: чоловіків серед звичайних (Normal) покупців більше, ніж жінок (90 проти 77).
- Yangon має більш збалансовану картину, без вираженого перекосу.

- Різниця може бути наслідком різних локальних маркетингових підходів — де краще працює залучення жінок до програми, а де — ні.
- Жінки в Naypyitaw можливо отримують більше користі або інформації про переваги програми (наприклад, через канали комунікації).
- Чоловіки в Mandalay можуть бути менш охочі доєднуватися до лояльності через відсутність персоналізованих або релевантних переваг.

Рекомендації:

- посилити комунікацію переваг програми лояльності, особливо в Mandalay, орієнтуючись на жінок.

- для чоловіків створити окремі тригери мотивації:

- спеціальні «чоловічі» пропозиції (знижки на електроніку, доглядові засоби, тощо).
Чітка проста програма: наприклад, кешбек без накопичень, «1 покупка = 1 бонус».
- ексклюзивні акції лише для учасників програми:

- «Таємний день знижок», спеціальні продукти або ціни лише для лояльних клієнтів.
- аналіз локального маркетингу в Naypyitaw — можливо, там ефективніше працюють точки впливу, які варто масштабувати на інші регіони.

In [ ]:
pd.pivot_table(df,index='Product line', columns=['Payment', 'Gender'], values='Quantity', aggfunc='sum')

Найбільше покупок за готівку (Cash) здійснюють у категорії Sports and travel, де загальна кількість покупок становить 324. Для Food and beverages переважає оплата кредитною карткою (Credit card) — 335 покупок. А от Home and lifestyle найчастіше купують через електронний гаманець (Ewallet) — 346 покупок. Це може свідчити про зручність цифрових оплат у категорії товарів для дому або акційні пропозиції, пов’язані з Ewallet.


- жінки найактивніше купують Fashion accessories та Home and lifestyle через електронний гаманець, а також використовують готівку при покупках Food and beverages і Sports and travel.
- чоловіки найбільше витрачають на Health and beauty через Ewallet, що може свідчити про або високий попит на чоловічу доглядову продукцію, або про покупки подарунків.
- для Electronic accessories жінки частіше платять готівкою, а чоловіки — електронним гаманцем.

####Висновки:
- жінки частіше використовують готівку, особливо в повсякденних категоріях.
- чоловіки активніші в категорії Health and beauty, і оплачують Ewallet — можливо, тут працюють персональні акції або рекомендації.
- Ewallet дуже популярний серед жінок для Fashion accessories та Home and lifestyle — потенційна зона для розвитку таргетованих акцій.

####Рекомендації для бізнесу:
- запровадити цільові акції за способом оплати — наприклад, знижки на Home and lifestyle при оплаті через Ewallet.
- сегментувати пропозиції за гендером та способом оплати: жінкам — візуально привабливі акції в категоріях Fashion та Home, чоловікам — чіткі пропозиції в Health and beauty.
- вивчити кампанії лояльності для Home and lifestyle, які працюють краще з Ewallet — і масштабувати їх на інші категорії.